### ISSUE TITLE
Dynamic linking of BLAS / LAPACK

### ISSUE THREAD

**Issue by rth** [iodide-project/pyodide#240]
This was discussed in https://github.com/iodide-project/pyodide/pull/211 I would like to reformulate the issue so it's easier for someone to continue working on it.

Scipy requires BLAS and LAPACK to build. Note: my earlier comments in https://github.com/iodide-project/pyodide/pull/211#issuecomment-428928156 about LAPACK being optional due to the note at the and of [this paragraph in scipy docs](https://scipy.github.io/devdocs/building/linux.html#generic-instructions) were actually not true, as was confirmed on [the scipy-dev mailing list](https://mail.python.org/pipermail/scipy-dev/2018-November/023157.html).

Currently in https://github.com/iodide-project/pyodide/pull/211 we link LAPACK statically, which means that it gets included in around ~8 compiled C extensions , with a resulting size of 170MB lz4 compressed / 650MB uncompressed. Different compile flags where tried in https://github.com/iodide-project/pyodide/pull/211 but ultimately it doesn't improve size much.

The most clean way to improve the package size would be to link dynamically, which would also allow to get rid of some hacks in `pyodide_build/pywasmcross.py` (https://github.com/iodide-project/pyodide/pull/211#discussion_r230204118). More information about dynamic linking in emscripten can be found in https://github.com/kripken/emscripten/wiki/Linking

Some of this was tried in https://github.com/iodide-project/pyodide/pull/211#discussion_r230204118 but didn't succeed. In retrospective, disabling validation of output in emcc / asm2wasm could be a first step. Then one would need to manually open the needed modules with `dlopen`, since `ldd` does not exist in emscripten (https://github.com/iodide-project/pyodide/pull/211#issuecomment-431966647).

**Comment by navytux** [comment]
Some feedback about dynamic linking from wasm/emscripten newbie discovered while learning it:

It is possible in emscripten to do dynamic linking not only from main module to other libraries, but from those other libraries to next other dynamic modules. (In scipy case the main module is main pyiodide, a DSO is a `*.so` from scipy, and those DSO want to link to e.g. `liblapack.so`). The way it works is that a `Module` in emscripten can have `.dynamicLibraries`, and then the runtime automatically loads referenced libraries. Today this works for main module already (one uses e.g. `-s RUNTIME_LINKED_LIBS="..."`) but does not automatically work from sided DSO to another DSO. However even today it could be worked around if DSO A wants to link to DSO B e.g. via

```c
// (liba.c)
__attribute__((constructor))
static void init() {
   // XXX EM_ASM does not work in wasm mode
#  if 1
   EM_ASM({
       loadDynamicLibrary('libB_so.js');
   });
#  else
     emscripten_run_script("loadDynamicLibrary('libB_so.wasm')");
#  endif
}
```

this just calls `loadDynamicLibrary` from emscripten runtime to do the job.

It works but is not good for real usage, as if e.g. A1 and A2 both would want to load B, then B would be loaded twice. It seems to be not too hard to fix in emscripten runtime however - please see notes here:

https://lab.nexedi.com/kirr/emscripten/commit/98006640

This way conventional support for shared libraries could be implemented in emscripten - similar to what we are used to have e.g. on Linux and other OS'es.

I've put a full test/demo for such linking here:

https://lab.nexedi.com/kirr/wasm-study  (rev `ee35d772` as of this writing).

Kirill

**Comment by navytux** [comment]
So I've implemented emscripten support for linking to shared libraries which in turn are linked to other shared libraries:

https://github.com/kripken/emscripten/pull/7512

This way all `*.so` scipy modules should be able to link to `lapack.so` dynamically.
@rth, could you please check whether it helps for scipy/lapack case?

@mdboom, I think this have the potential to also remove freetype and libpng  from main pyodide, since they are only used by `matplotplib.so` module (if I understood correctly).

Thanks,
Kirill

**Comment by rth** [comment]
Very impressive work @navytux ! Thank you!

For scipy, once https://github.com/iodide-project/pyodide/pull/211 is merged I think we can make another PR where we apply your patches to emscripten and see whether this solves the problem.

**Comment by navytux** [comment]
@rth, thanks for feedback.

Just for the reference: the build system of scipy will have to be adjusted to leverage new library -> library linking feature. I mean it won't work automatically with just emscripten ldso patch and current unchanged way to build scipy.

**Comment by navytux** [comment]
My patches for DSO -> DSO linking landed into Emscripten incoming and should be hopefully available as part of upcoming Emscripten 1.38.22 release.

@rth, it would be interesting to see how it helps scipy packaging. You can see how to use the linking e.g. here: https://lab.nexedi.com/kirr/wasm-study/blob/de232252f8/Makefile#L34.

@mdboom, I also have the feeling that after the [fix](https://github.com/kripken/emscripten/commit/8ec4b4ab) for concurrent loading of wasm modules, preloading them serially ([1](https://github.com/kripken/emscripten/blob/1.38.21-27-g9ecca00df/src/library_browser.js#L242), [2](https://github.com/kripken/emscripten/commit/a5866a50)) is no longer required. I've actually had to do the fix while testing concurrent DSO loading under browser environment: https://github.com/kripken/emscripten/blob/1.38.21-27-g9ecca00df/tests/runner.py#L799, and this way it should for sure help preloading too. But I had not tried myself.

Kirill

**Comment by mdboom** [comment]
Thanks for all this great work!

I created #278 to remind us to do the emscripten upgrade when it's ready.  Then we can come back to this issue and rework the Scipy build to use dynamic linking against BLAS and watch our file sizes shrink.

I've also created #279 to remind us to come back and look into the serial wasm module loading later.

**Comment by rth** [comment]
Thanks a lot for your hard work on this @navytux ! I'm really glad DSO -> DSO linking works in Emscritpen now.

Yes, as soon as Emscripten 1.38.22 is out and we migrate to it, we'll have a look at scipy, and will keep you informed of how it goes.

**Comment by navytux** [comment]
@mdboom, @rth, thanks for feedback and kind words. Good luck with the tries.

**Comment by navytux** [comment]
@mdboom, @rth, any news?

**Comment by rth** [comment]
@navytux It's on my todo list, will try to get to it on the time scale of one month or two, unless someone does it before.

Thanks again for your work on this!

**Comment by navytux** [comment]
@rth, thanks for feedback. Good luck there.

**Comment by manumartin** [comment]
Hi @navytux I'm trying to build scipy with dynamic linking using your DSO->DSO feature on emscripten that was published on emscripten 1.38.22. Do you know the state of this change? I'm trying to make a small POC work first but I'm not able to do it. I have tried to run your POC "wasm-study" and it doesn't seem to work on emscripten 1.38.47. The build looks like this:

```
emcc -s WASM=1 -s SIDE_MODULE=1 -DLIBF_COMPILE -o wasm/libf_so.wasm libf.c
emcc -s WASM=1 -s SIDE_MODULE=1 -s RUNTIME_LINKED_LIBS="['libf_so.wasm']" -DLIBK_COMPILE -o wasm/libk_so.wasm libk.c
emcc -s WASM=1 -s MAIN_MODULE=1 -s RUNTIME_LINKED_LIBS="['libk_so.wasm']" -o wasm/prog.js prog.c
```
which looks correct to me. When I execute it it manages to resolve kfunc from the first module but then it can't resolve ffunc from the second module and raises the following error:

```
Assertion failed: missing linked function `_ffunc1`. perhaps a side module was not linked in? if this global was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment
```
From reading the code and if I understood it well: you made a JS proxy that gets executed on every function call that the first SIDE_MODULE makes, this proxy is what should load the second library on demand when it finds an undefined symbol, so the process should be:

- _kfunc1 stub in main.js is executed
- kfunc1 inside libk_so.wasm is executed
- the JS proxy in main.js is executed and resolvSymbol determines no symbol for ffunc1 exists
- the JS proxy reads the dylink section on libk_so.wasm and determines libf_so.wasm must be loaded and loads it.
- _ffunc1 is loaded into Module
- _ffunc1 is called

However resolveSymbol isn't doing anything more, it just sees that Module['_ffunc1] is undefined and throws an error and that's it.

Can you please help me or give me some directions on how to make this work?
thanks!

**Comment by manumartin** [comment]
@navytux I'm fairly confused.. I'm seeing that the only place where WebAssembly.make_shared_library is called inside emcc.py is in the do_binaryen function.
Since I'm working with wasm and LLVM the WASM_BACKEND flag is enabled, but the call is being prevented by this check:
```
  if shared.Settings.SIDE_MODULE and not shared.Settings.WASM_BACKEND:
    wso = shared.WebAssembly.make_shared_library(final, wasm_binary_target, shared.Settings.RUNTIME_LINKED_LIBS)
```
Does this mean DSO's are not currently supported with WASM/LLVM? only for asm.js?
I'm going to ask to the emscripten people.

**Comment by manumartin** [comment]
I think I understand this now, when the LLVM backend is activated the task of building the .wasm object is delegated to LLVM. This make_shared_library code seems to be there to work with the older backend. When using LLVM, LLVM should be in charge of building the .wasm file and generating the proper dylink section.

**Comment by navytux** [comment]
@manumartin, thanks for asking.

I had not touched emscripten for a long time and right now I do not have its checkout ready to play. For direction I suggest to look at my changes merged at https://github.com/emscripten-core/emscripten/pull/7512 (see the end of the first message - it references all steps that have been merged). Those patches were adding tests for dynamic linking to emscripten. I believe those tests should be definitive guide on how to use DSO -> DSO dynamic linking, as even if emscripten user interface/flags changed, those test should have been updated as well. At least that's how I would investigate.

Hope it helps a bit,
Kirill

**Comment by rth** [comment]
To respond to this [Gitter comment](https://gitter.im/iodide-project/iodide?at=5db0572de886fb5aa2dfce95) by @manumartin,
> I think dso->dso linking is only supported with the old backend but not with LLVM
llvm is capable of generating .wasm objects with the needed dylink section, and emscripten knows how to dynamically load those objects by reading that section

Good to know! I though that we needed to migrate backends because in https://github.com/iodide-project/pyodide/pull/473 I was blocked by https://github.com/WebAssembly/binaryen/issues/2177#issuecomment-504269479 and it was suggested that updating to llvm-upstream could be a way around it. I'm not saying that we can't make it work with the current backend, but there is also a question that if that migration is bound to happen at some point, then applying fixes for fascomp backend might not be the best way to spend efforts, at least in the medium term.

**Comment by manumartin** [comment]
Hi @rth, right now I'm not sure if "transitive" dynamic linking is supported or not by the fastcomp backend but I'm trying to build with LLVM anyway because of other potential benefits. It seems there is some amount of support, for example the .js loader generated by emscripten is able to read .wasm dylink sections in order to know which shared dependencies it needs to load. LLVM also has support for generating the section, in fact the SIDE_MODULE's that emcc generates with LLVM contain the section, it just happens to be empty. My guess is that emcc.py should at some point give LLVM the dependency information so that it can fill that section and with that it should work.

I managed to build cpython with LLVM and I'll try to commit those changes as soon as possible but it isn't currently working because of runtime errors that I think are related to -fPIC not working well with LLVM.

In fact the -fPIC runtime issues are happening with a very simple hello world c program and I think people from LLVM are aware of it . I have compared and debugged generated WASM with both -fPIC and without it and the -fPIC version seems to be having issues with references to things like static strings.

**Comment by rth** [comment]
With the upstream LLVM backend in emscripten https://github.com/iodide-project/pyodide/pull/1102 scipy with statically linked BLAS/LAPACK is now 35MB instead of 165MB (uncompressed) before. So it looks like elimination of unused objects works much better at link time with latest emscripten. We can still explore dynamic linking but it's is now much less critical. Among other things scipy should now be cached by the browser by default.
